### ISSUE TITLE
CI: Cilium E2E Upgrade: `Timed out waiting for datapath updates of FQDN IP information` after upgrade

### ISSUE THREAD

**Issue by pchaigno** [cilium/cilium#29846]
From https://github.com/cilium/cilium/actions/runs/7190938554/job/19584876380
```
[=] Test [check-log-errors] [67/67]

  [-] Scenario [check-log-errors/no-errors-in-logs]
  Found "2023-12-13T05:22:11.945479031Z level=error msg=\"Timed out waiting for datapath updates of FQDN IP information; returning response\" subsys=daemon" in logs 1 times

  ❌ Found 1 logs matching list of errors that must be investigated:
2023-12-13T05:22:11.945479031Z level=error msg="Timed out waiting for datapath updates of FQDN IP information; returning response" subsys=daemon
```
Sysdump: [cilium-sysdump-5-20231213-052359.7z.zip](https://github.com/cilium/cilium/files/13659250/cilium-sysdump-5-20231213-052359.7z.zip)

**Comment by mathpl** [comment]
Didn't see it in dogfooding hence I closed https://github.com/isovalent/customer-support/issues/557 which looks like the same issue, but still happening in CI.

**Comment by squeed** [comment]
Filed a PR should help reduce these timeouts significantly. Found a needless regeneration on the FQDN path that caused lock contention.

**Comment by giorio94** [comment]
Hit on Conformance GKE as well after enabling the check for log errors also there: https://github.com/cilium/cilium/actions/runs/7208580138/job/19637888894#step:21:182

**Comment by giorio94** [comment]
Reopening, as hit in the Cilium IPsec upgrade workflow for a PR which includes the commits from https://github.com/cilium/cilium/pull/29865.

* PR: https://github.com/cilium/cilium/pull/30012
* Link: https://github.com/cilium/cilium/actions/runs/7278734523/job/19833561516

**Comment by pippolo84** [comment]
Another hit [here](https://github.com/cilium/cilium/actions/runs/7384604115/job/20087835031).

```
[=] Test [check-log-errors] [71/71]

  [-] Scenario [check-log-errors/no-errors-in-logs]
  Found "2024-01-02T10:54:30.033766201Z level=error msg=\"Timed out waiting for datapath updates of FQDN IP information; returning response\" subsys=daemon" in logs 1 times

  ❌ Found 1 logs matching list of errors that must be investigated:
2024-01-02T10:54:30.033766201Z level=error msg="Timed out waiting for datapath updates of FQDN IP information; returning response" subsys=daemon
```

[Full logs](https://github.com/cilium/cilium/files/13820404/logs_1986399.zip)
Sysdump too big :(

**Comment by squeed** [comment]
Taking a look.

**Comment by squeed** [comment]
Aha, the backport to v1.15 hasn't merged yet. Whew.

**Comment by pippolo84** [comment]
> Taking a look.

Thanks! I kept a local copy of the sysdump, let me know if you want to take a look.

**Comment by squeed** [comment]
@pippolo84 are these failures on the backport PR that includes this change? GitHub is not making it easy to determine.

**Comment by pippolo84** [comment]
> @pippolo84 are these failures on the backport PR that includes this change? GitHub is not making it easy to determine.

Yep, the workflow has been triggered by [this PR](https://github.com/cilium/cilium/pull/30079).

**Comment by pippolo84** [comment]
Oh, you were looking for your changes from https://github.com/cilium/cilium/pull/29865. Those have been backported in [the PR](https://github.com/cilium/cilium/pull/30079) that triggered the workflow, so it seems the issue is still there, unfortunately.

**Comment by squeed** [comment]
So, some basic analysis:

- There are two practically-concurrent requests, for `A one.one.one.one.` and `AAAA one.one.one.one.`
- The timeout message comes from the second request, in this case, AAAA.
- The vast majority of that time is spent waiting for the ipcache to complete

The basic flow

10:54:29.915: Response to `A one.one.one.one.` is received
10:54:29.921: NameManager is updated; locks released, waiting for ipcache to process v4 addrs. ipcache immediately starts.
10:54:29.923: Response to `AAAA` is received.
10:54:29.927: NameManager is updated; locks released, waiting for ipcache to process v6 addrs
10:54:29.933: ipcache PolicyMap updates are complete, waiting for proxy (Envoy) updates
10:54:29.946: Proxy updates are complete, ipcache is complete, total duration ~24ms, identity allocation ~1ms, proxy update ~9ms
10:54:29.946: ipcache starts again.
10:54:29.947: DNS response is released for `A` request. Total time: 32ms
10:54:29.987: ipcache PolicyMap updates are complete, waiting for proxy (Envoy) updates
10:54:30.033: We give up waiting for ipcache, write DNS response back.
10:54:30.037: Proxy updates are complete, ipcache is complete, total duration 93ms, identity allocation ~20ms, proxy update ~50ms, identity allocation 15ms

So, two observations:
- Identity allocation randomly takes a long time. I'm assuming this is due to GC pauses / allocation
- Envoy can also take a long time.

I don't see any smoking guns; 20ms lost in a trivial map update and 50ms lost waiting for envoy, plus the rest of the FQDN process, put us over 100ms. I'll try and dig in to why this is going wrong.

Next step is to find out why.

**Comment by squeed** [comment]
Created a gist with a bit more information here: https://gist.github.com/squeed/752d105c569db3eb328de191d37a4ed8

I'll ask around for help.

**Comment by squeed** [comment]
I did some exploration with @jrajahalme's assistance, and I found that Envoy sometimes just takes 20-30ms to process updates. There are no obvious performance smoking guns.

I would like to consider bumping the timeout to 150ms in CI to see if that gets rid of flakes. All the flakes I saw were timing out in ~103ms, the threshold is 100ms. I think we're just too resource constrained here.

**Comment by squeed** [comment]
I've seen this error message hit on some v1.14 CI jobs, which *does not* have my FQDN refactor. So I suspect this has always been an issue. We may wish to consider ignoring this warning. Perhaps it is allowed to log once, but more than that is indicative of an issue. Hard to say; it is literally a threshold matter.

**Comment by pchaigno** [comment]
Couldn't users hit this as well? If so, ignoring in CI may not be enough.

FYI, this is still happening a lot in CI and currently preventing us from making some workflows Required.

**Comment by squeed** [comment]
Users can and do hit this in production, it has been this way for years. We don't have good data, though.

If we want to continue blocking CI, then we can keep this as we figure out how to redesign the Envoy system. Otherwise, we should skip this error message.

**Comment by pchaigno** [comment]
Sounds like we may need to allowlist in CI + document.

**Comment by squeed** [comment]
> Sounds like we may need to allowlist in CI + document.

Agreed, I'll take care of that shortly. And figure out how we can improve Envoy.

**Comment by tommyp1ckles** [comment]
@squeed Created a[ PR on cilium-cli ](https://github.com/cilium/cilium-cli/pull/2372) to add this to the list of exceptions. Can I get your confirmation that this is the appropriate way to handle this?

**Comment by pchaigno** [comment]
This is also affecting at least Conformance GKE: https://github.com/cilium/cilium/actions/runs/8556001839/job/23444773987.

@squeed If you don't have time to implement the fix, could you unassign yourself and ask for someone else in your team to handle it? :pray:

**Comment by squeed** [comment]
Filed https://github.com/cilium/cilium/pull/31866 to bump the timeout to 250ms. There have also been a few FQDN performance improvements:

- #31454, merged Apr 3
- #30897, merged Apr 3

Hopefully this cuts the noise down significantly.

**Comment by pchaigno** [comment]
The GKE occurrence above is from April 4. This one as well: https://github.com/cilium/cilium/actions/runs/8558512353/job/23453228166. Both on main. It may have reduced it (?) but it doesn't look like it was enough.

**Comment by pchaigno** [comment]
This is still happening on main: https://github.com/cilium/cilium/actions/runs/8690522661/job/23830675920.

**Comment by giorio94** [comment]
Just hit on v1.14 as well: https://github.com/cilium/cilium/actions/runs/8703624933/job/23870214379

**Comment by squeed** [comment]
Requested backport to v1.15 and v1.14.

**Comment by julianwiedmann** [comment]
Haven't seen this in a long time. Unassigning so that stale-issue GC can take over.
